File: examples/lib/time/spacetime/spacetime.php
<?php
/**
 * spacetime - Format and display current time with optional template support
 *
 * Usage: spacetime([template])
 *   Without arguments: Returns formatted time as
 *     "DayOfWeek YYYY-MM-DD HH:MM:SS TZ Timezone"
 *   With template: Returns custom format using placeholders:
 *     {{dow}}      - Day of week (e.g., Monday)
 *     {{date}}     - Date in YYYY-MM-DD format
 *     {{time}}     - Time in HH:MM:SS format
 *     {{tz}}       - Timezone offset (e.g., +0000)
 *     {{timezone}} - Timezone name (e.g., America/New_York)
 *
 * @param string|null $template Optional template string with placeholders
 * @return string Formatted time string
 */
function spacetime(?string $template = null): string {
  // Resolve timezone name
  // If TZ env var is set (even empty), use it directly (POSIX: TZ='' means UTC).
  // If TZ is unset, use cached system timezone.
  static $cached_timezone = null;
  $tz_env = getenv('TZ');
  if ($tz_env !== false) {
    $timezone_name = ($tz_env === '') ? 'UTC' : $tz_env;
    try {
      $tz_obj = new DateTimeZone($timezone_name);
    } catch (\Exception $e) {
      $timezone_name = 'UTC';
      $tz_obj = new DateTimeZone('UTC');
    }
  } else {
    if ($cached_timezone === null) {
      $cached_timezone = date_default_timezone_get();
    }
    $timezone_name = $cached_timezone;
    $tz_obj = new DateTimeZone($timezone_name);
  }

  $now = new DateTime('now', $tz_obj);
  $dow = $now->format('l');
  $date = $now->format('Y-m-d');
  $time = $now->format('H:i:s');
  $tz = $now->format('O'); // +0000 format (no colon)
  $timezone = $timezone_name;

  // Default format when no template provided
  $curtime = "$dow $date $time $tz $timezone";

  // If a template string is provided, parse and replace placeholders
  if ($template !== null) {
    $curtime = str_replace(
      ['{{dow}}', '{{date}}', '{{time}}', '{{tz}}', '{{timezone}}'],
      [$dow, $date, $time, $tz, $timezone],
      $template
    );
  }

  // Store the result in the global variable for later reference
  $GLOBALS['EPOCHSPACETIME'] = $curtime;

  return $curtime;
}

// If script is executed directly, run spacetime with arguments
if (basename(__FILE__) === basename($_SERVER['SCRIPT_FILENAME'] ?? '')) {
  $VERSION = '1.1.0';

  if (isset($argv[1]) && in_array($argv[1], ['-V', '--version'])) {
    echo basename(__FILE__) . " $VERSION\n";
    exit(0);
  }

  if (isset($argv[1]) && in_array($argv[1], ['-h', '--help'])) {
    echo <<<'EOF'
spacetime - Format and display current time with template support

Usage: spacetime [template]

Without arguments: Returns formatted time as
  "DayOfWeek YYYY-MM-DD HH:MM:SS TZ Timezone"

With template: Returns custom format using placeholders

Placeholders:
  {{dow}}      - Day of week (e.g., Monday)
  {{date}}     - Date in YYYY-MM-DD format
  {{time}}     - Time in HH:MM:SS format
  {{tz}}       - Timezone offset (e.g., +0000)
  {{timezone}} - Timezone name (e.g., America/New_York)

Examples:
  spacetime.php
  TZ=UTC spacetime.php
  spacetime.php "{{date}} at {{time}}"
  spacetime.php "Log entry: {{dow}} {{date}} {{time}} {{tz}}"
EOF;
    echo "\n";
    exit(0);
  }

  echo spacetime($argv[1] ?? null) . "\n";
}

#fin

File: examples/lib/time/spacetime/spacetime.py
#!/usr/bin/env python3
"""spacetime - Format and display current time with optional template support.

Usage: spacetime([template])
  Without arguments: Returns formatted time as
    "DayOfWeek YYYY-MM-DD HH:MM:SS TZ Timezone"
  With template: Returns custom format using placeholders:
    {{dow}}      - Day of week (e.g., Monday)
    {{date}}     - Date in YYYY-MM-DD format
    {{time}}     - Time in HH:MM:SS format
    {{tz}}       - Timezone offset (e.g., +0000)
    {{timezone}} - Timezone name (e.g., America/New_York)
"""
from __future__ import annotations

import os
import sys
from datetime import datetime
from zoneinfo import ZoneInfo, ZoneInfoNotFoundError

__version__ = '1.1.0'
__all__ = ['spacetime', 'EPOCHSPACETIME', '__version__']

EPOCHSPACETIME: str = ''
_cached_timezone: str | None = None


def _resolve_system_timezone() -> str:
  """Resolve the system timezone name via a fallback chain.

  Mirrors the bash implementation:
    1. /etc/timezone (standard on Debian/Ubuntu)
    2. readlink /etc/localtime target (zoneinfo path)
    3. UTC
  """
  try:
    with open('/etc/timezone', encoding='utf-8') as f:
      name = f.read().strip()
      if name:
        return name
  except OSError:
    pass
  try:
    link = os.readlink('/etc/localtime')
    marker = 'zoneinfo/'
    idx = link.find(marker)
    if idx != -1:
      return link[idx + len(marker):]
  except OSError:
    pass
  return 'UTC'


def spacetime(template: str | None = None) -> str:
  """Format current time, optionally via a placeholder template.

  Honours TZ env var. Per POSIX, TZ='' (set but empty) means UTC.
  When TZ is unset, the system timezone is resolved once and cached.
  The result is also stored in the module-level EPOCHSPACETIME.
  """
  global EPOCHSPACETIME, _cached_timezone

  # Resolve timezone: honour TZ env var (set-but-empty → UTC per POSIX).
  if 'TZ' in os.environ:
    tz_name = os.environ['TZ'] or 'UTC'
  else:
    if _cached_timezone is None:
      _cached_timezone = _resolve_system_timezone()
    tz_name = _cached_timezone

  try:
    tz_obj = ZoneInfo(tz_name)
  except ZoneInfoNotFoundError:
    tz_name = 'UTC'
    tz_obj = ZoneInfo('UTC')

  now = datetime.now(tz_obj)
  dow = now.strftime('%A')
  cur_date = now.strftime('%Y-%m-%d')
  cur_time = now.strftime('%H:%M:%S')
  tz = now.strftime('%z')  # +0700 format (no colon)
  timezone = tz_name

  if template is None:
    result = f'{dow} {cur_date} {cur_time} {tz} {timezone}'
  else:
    result = (
      template
      .replace('{{dow}}', dow)
      .replace('{{date}}', cur_date)
      .replace('{{time}}', cur_time)
      .replace('{{tz}}', tz)
      .replace('{{timezone}}', timezone)
    )

  EPOCHSPACETIME = result
  return result


_HELP_TEMPLATE = '''\
__NAME__ __VERSION__ - Format and display current time with template support

Usage: __NAME__ [template]

Without arguments: Returns formatted time as
  "DayOfWeek YYYY-MM-DD HH:MM:SS TZ Timezone"

With template: Returns custom format using placeholders

Placeholders:
  {{dow}}      - Day of week (e.g., Monday)
  {{date}}     - Date in YYYY-MM-DD format
  {{time}}     - Time in HH:MM:SS format
  {{tz}}       - Timezone offset (e.g., +0000)
  {{timezone}} - Timezone name (e.g., America/New_York)

Examples:
  __NAME__
  TZ=UTC __NAME__
  __NAME__ "{{date}} at {{time}}"
  __NAME__ "Log entry: {{dow}} {{date}} {{time}} {{tz}}"'''


def _main(argv: list[str]) -> int:
  script_name = os.path.basename(argv[0]) if argv else 'spacetime.py'
  args = argv[1:]

  if args and args[0] in ('-V', '--version'):
    print(f'{script_name} {__version__}')
    return 0

  if args and args[0] in ('-h', '--help'):
    print(_HELP_TEMPLATE.replace('__NAME__', script_name).replace('__VERSION__', __version__))
    return 0

  template = args[0] if args else None
  print(spacetime(template))
  return 0


if __name__ == '__main__':
  sys.exit(_main(sys.argv))

#fin
